Add remaining sections to chap. 2 and update links (#19)
wso2/container-guide: push/Push.md

The following will be discussed in detail in the document.

* [2.1. Why should you have your own private registry ?](#hhy-private-registry)
* [2.2. How to secure your private registry ?](#how-to-secure)
* [2.3. Scanning and Auditing](#scanning-auditing)

<br/>

> ## <a name="why-private-registry"></a>2.1. Why should you have your own private registry ?

Reviewer: ```suggestion
## <a name="why-private-registry"></a>2.1. Why should you have your own private container registry ?
```

Author: done

> This is essential due to a couple of reasons:

Reviewer: ```suggestion
It is salient to own your private container registry because of the following reasons:
```

Author: done

This is also applicable to product Docker images available at [WSO2 Private Docker Registry](https://docker.wso2.com) mainly due to limitations of storage. For each product version, new images are being frequently added with the latest [WSO2 Updates](https://wso2.com/updates) to products and changes to docker resources used. With limitations of storage, not all images being pushed could be retained in the registry for a long time.

> As a result, WSO2 best advise the use of your own private container registry to host all the required images in your deployments, so that you have full control over them. In a rollback event, the recovery of previous version images is effortless with zero risks and zero dependence on outside sources.

Reviewer 2: Would it be better to mention that the private registry should be a HA deployment since the application deployment process depends on it?.

Author: @ThilinaManamgoda No, let's not mention that here since it is an implementation level detail.

> As a result, WSO2 best advise the use of your own private container registry to host all the required images in your deployments, so that you have full control over them. In a rollback event, the recovery of previous version images is effortless with zero risks and zero dependence on outside sources.

Reviewer: ```suggestion
As a result, WSO2 best advise the use of your own private container registry to host all the required images in your deployments, so that you have full control over them. In a rollback event, the recovery of previous version images is effortless with zero risks and dependence on outside sources.
```

Author: done

> ## <a name="how-to-secure"></a>2.2. How to secure your private registry ?

Reviewer: ```suggestion
## <a name="how-to-secure"></a>2.2. How to secure your private  container registry ?
```

Author: done

> Your private registry is the place where you host all the images required for your deployment. When you push an image to the registry, it is always important to be sure that you push the exact image that you want to store in your registry. When you pull an image from the registry, it is also important to be sure that you pull the exact image that you want to retrieve from your registry. Since the transfer of data in both of these situations happens in HTTP, to avoid any man-in-the-middle attacks, securing the transfer using `SSL security` is essential. 

Reviewer 3: @DilanUA shall we break the `Since the transfer of data in both of these situations happens in HTTP, to avoid any man-in-the-middle attacks, securing the transfer...` sentence into a separate paragraph or point?

Author: @chirangaalwis How about this? Since the transfer of data in both of these situations happens in HTTP, securing the transfer using `SSL security` is essential to avoid any man-in-the-middle attacks.

Reviewer: +1

Author: done

> Your private registry is the place where you host all the images required for your deployment. When you push an image to the registry, it is always important to be sure that you push the exact image that you want to store in your registry. When you pull an image from the registry, it is also important to be sure that you pull the exact image that you want to retrieve from your registry. Since the transfer of data in both of these situations happens in HTTP, to avoid any man-in-the-middle attacks, securing the transfer using `SSL security` is essential. 

Reviewer: ```suggestion
Your private Container registry is the place where you host all the images required for your deployment. When you push an image to the registry, it is always important to be sure that you push the exact image that you want to store in your registry. On the contrary, when an image is pulled from the registry, it is vital to be sure that you pull the correct image that you want to retrieve from your registry. 
```

> Secondly to provide access only to authorized personel even with in your organization, it is important that you enable atleast `basic authentication` to secure your registry.  

Reviewer: ```suggestion
Secondly to provide access only to authorized personnel even within your organization, it is important that you enable at least `basic authentication` to secure your registry.  
```
Shall we ommit the Secondly word as there is no point with First in the text? Instead, we can your Further,
Within simply the correct ways of writing this word.  Within not be written with a space between with/in or with/out, according to all major dictionaries

Author: done

## <a name="scanning-auditing"></a>2.3. Scanning and Auditing

> If an image includes a vulnerability, then every container generated using that image will also include that vulnerability. New images and newer versions of existing images may be pushed frequently into your private registry. Every change carries with it the possibility of introducing a new vulnerability - and all too often, an old vulnerability resulting from the use of an outdated software package.

Reviewer: ```suggestion
If an image includes a vulnerability, then every container generated using that particular image will also include vulnerability. New images and newer versions of existing images should be pushed periodically into your private registry. Every change carries with it the possibility of introducing a new vulnerability and all too often, an old vulnerability resulting from the use of an outdated software package.
```

Author: done

> `Scanning` : You need to schedule regular scans of your repository in order to detect such vulnerabilities. Most of the organizations that maintain major public registries (such as Docker) provide their own scanning services. Even for private registries, there are a plenty of security providers to offer extended scanning capabilities, with key features such as scanning-results alerts, and automatically preventing images that are not trusted from being used.

Reviewer 3: @DilanUA what is meant here by `Most of the organizations that maintain major public registries (such as Docker) provide their own scanning services.`?

Shouldn't it be major private registries? External organizations do not have any control over popular public registry (e.g. DockerHub, Quay Container Registry) scanning procedures AFAIK.

Author: This means even the public registry maintainers do have scanning as a best practice.

Reviewer 3: @DilanUA if so I think it has to be `DockerHub` rather than `Docker`.

Author: changing to: Most of the organizations that maintain major public registries provide their own scanning services (such as [Docker](https://docs.docker.com/docker-hub/official_images/#official-image-vulnerability-scanning) for their official images).

> `Scanning` : You need to schedule regular scans of your repository in order to detect such vulnerabilities. Most of the organizations that maintain major public registries provide their own scanning services (such as [Docker](https://docs.docker.com/docker-hub/official_images/#official-image-vulnerability-scanning) for their official images). Even for private registries, there are a plenty of security providers to offer extended scanning capabilities, with key features such as scanning-results alerts, and automatically preventing images that are not trusted from being used.

Reviewer 2: We can point to https://goharbor.io/ as an example private registry

> `Scanning` : You need to schedule regular scans of your repository in order to detect such vulnerabilities. Most of the organizations that maintain major public registries provide their own scanning services (such as [Docker](https://docs.docker.com/docker-hub/official_images/#official-image-vulnerability-scanning) for their official images). Even for private registries, there are a plenty of security providers to offer extended scanning capabilities, with key features such as scanning-results alerts, and automatically preventing images that are not trusted from being used.

Reviewer: ```suggestion
`Scanning` : You need to schedule regular scans of your repository in order to detect vulnerabilities. Most of the organizations that maintain major public registries provide their own scanning services (such as [Docker](https://docs.docker.com/docker-hub/official_images/#official-image-vulnerability-scanning) for their official images). Even for private registries, there are plenty of security providers to extending their expertise in scanning capabilities, with key features such as scanning-results alerts, and automatically preventing images that are not trusted from being used.
```

> `Auditing` : You can (and should) go beyond scanning for vulnerabilities by auditing images for age, as well as outdated packages. You can flag older images and those with older dependencies as outdated and schedule them for updating. Adopting a policy of regularly updating and refreshing older images makes it easier to eliminate vulnerabilities which may have gone undetected.

Reviewer: ```suggestion
`Auditing` : You can (and should) go beyond scanning for vulnerabilities by auditing images for age, as well as outdated packages. You can flag older images and those with older dependencies as outdated and schedule them for updating. Adopting a policy of regularly updating and refreshing older images makes it easy to eliminate vulnerabilities that may have gone undetected.
```

Author: done
